asn1ate/asn2quickder.py

```diff
@@ -0,0 +1,321 @@
+	"""Replace unsupported characters in ASN.1 symbol names"""
+	return str (name).replace (' ', '').replace ('-', '_')
+
+
+class QuickDERgen ():
+	"""Generate the C header files for Quick DER, a.k.a. Quick and Easy DER.
+
+	   There are two things that are generated for each of the ASN.1 syntax
+	   declaration symbol of a unit:
+
+	   #define DER_PACK_unit_SyntaxDeclSym \
+			   DER_PACK_ENTER | ..., \
+			   ... \
+			   DER_PACK_LEAVE, \
+			   DER_PACK_END
+
+	   this is a walking path for the der_pack() and der_unpack() instructions.
+	   In addition, there will be a struct for each of the symbols:
+
+	   struct unit_SyntaxDeclSym_ovly {
+		   dercursor field1;
+		   dercursor field2;
+		   struct unit_EmbeddedSym_ovly field3;
+		   dercursor field4;
+	   };
+
+	   The unit prefix will be set to the filename of the module, usually
+	   something like rfc5280 when the parsed file is rfc5280.asn1 and the
+	   output is then written to rfc5280.h for easy inclusion by the C code.
+	"""
+
+	def __init__ (ik, semamod, outfn, refmods):
+		ik.semamod = semamod
+		ik.refmods = refmods
+		if outfn [-2:] == '.h':
+			raise Exception ('File cannot overwrite itself -- use another extension than .h for input files')
+		ik.unit = toCsym (outfn.rsplit ('.', 1) [0])
+		ik.outfile = open (ik.unit + '.h', 'w')
+		ik.wout = ik.outfile.write
+		# Setup function maps
+		ik.ovly_funmap = {
+			DefinedType: ik.ovlyDefinedType,
+			ValueAssignment: ik.ignore_node,
+			TypeAssignment: ik.ovlyTypeAssignment,
+			TaggedType: ik.ovlyTaggedType,
+			SimpleType: ik.ovlySimpleType,
+			BitStringType: ik.ovlySimpleType,
+			ValueListType: ik.ovlySimpleType,
+			SequenceType: ik.ovlyConstructedType,
+			SetType: ik.ovlyConstructedType,
+			ChoiceType: ik.ovlyConstructedType,
+			SequenceOfType: ik.ovlySimpleType,	# var sized
+			SetOfType: ik.ovlySimpleType,		# var sized
+			ComponentType: ik.ovlySimpleType,  #TODO#
+		}
+		ik.pack_funmap = {
+			DefinedType: ik.packDefinedType,
+			ValueAssignment: ik.ignore_node,
+			TypeAssignment: ik.packTypeAssignment,
+			TaggedType: ik.packTaggedType,
+			SimpleType: ik.packSimpleType,
+			BitStringType: ik.packSimpleType,
+			ValueListType: ik.ovlySimpleType,
+			SequenceType: ik.packSequenceType,
+			SetType: ik.packSetType,
+			ChoiceType: ik.packChoiceType,
+			SequenceOfType: ik.packSequenceOfType,
+			SetOfType: ik.packSetOfType,
+			ComponentType: ik.packSimpleType,  #TODO#
+		}
+
+	def newcomma (ik, comma, firstcomma=''):
+		ik.comma0 = firstcomma
+		ik.comma1 = comma
+
+	def comma (ik):
+		ik.wout (ik.comma0)
+		ik.comma0 = ik.comma1
+
+	def close (ik):
+		ik.outfile.close ()
+
+	def generate_head (ik):
+		ik.wout ('/*\n * asn2quickder output for ' + ik.semamod.name + ' -- automatically generated\n *\n * For information on Quick `n\' Easy DER, see https://github.com/vanrein/quick-der\n *\n * For information on the code generator, see https://github.com/vanrein/asn2quickder\n *\n */\n\n\n#include <quick-der/api.h>\n\n\n')
```

Comment:
It would be nice if you added a `def writeline` method that writes a string followed by a newline.

Then phrase this as something like:

     self.writeline('/*')
     self.writeline(' * asn2quickder output for ' + ik.semamod.name + ' -- automatically generated')
     self.writeline(' *')
     self.writeline(' * For information on [...], etc

That makes it much easier to read than with embedded `\n` characters.